### ISSUE TITLE
[C] syntax assumes struct/enum/union keywords are always a "class" declaration

### ISSUE THREAD

**Issue by speeder** [sublimehq/Packages#716]
I decided for some random reason on my most recent project, to not use the "typedef" trick.

So for example when I will use a struct on my code, I type it like this:

```c
struct UI_BoundingBox decorativeBox = {10, titleHeight-3, width-20, height-10};
```

or this:

```c
struct UI_MenuBoxData
{
    struct UI_BoundingBox position;
    enum UI_BoxCharType borderType;
    unsigned int paddingX;
    unsigned int paddingY;
    struct UI_ScrollBoxText boxContents[];
};
```

The indexer, highlighter, etc... all assume that for example my UI_BoxCharType is creating a new type of enum, as result "go to definition" dumps me with a long list with every single place I used "enum UI_BoxChartype" (instead of using a typedef version of this).

**Comment by wbond** [comment]
There are a number of tests that prevent this from happening when using `struct mystruct myfunc() { }` and similar constructs. It seems we missed checking for these when parsing member structures.

It will be possible that sometimes we can't detect these properly, since the lexer is currently only able to lex a single line at a time. However, as long as code is written in the style above, e.g. `struct mystruct member;`, it should be possible.

Here is where the code lives:
 
 - https://github.com/sublimehq/Packages/tree/master/C%2B%2B
 - https://github.com/sublimehq/Packages/tree/master/Objective-C
 - All changes to C must also go in C++, Objective-C and Objective-C++

Specifically, here are some places in the definition that are relevant:

 - https://github.com/sublimehq/Packages/blob/master/C%2B%2B/C.sublime-syntax#L15
 - https://github.com/sublimehq/Packages/blob/master/C%2B%2B/C.sublime-syntax#L446-L499
 - https://github.com/sublimehq/Packages/blob/master/C%2B%2B/C.sublime-syntax#L512-L516

This seems to be the crux of the issue:

 - https://github.com/sublimehq/Packages/blob/master/C%2B%2B/C.sublime-syntax#L514-L515

Instead of pushing into parsing a `struct`, `union`, `enum`, a new rule should be added right before that checking for a `struct`, `union`, or `enum` followed by two identifiers and a `;`.

The scope names can be derived by looking at other parts of the syntax. Tests will also need to be added for all four languages. Look for the files named `syntax_test_*` in the C++ and Objective-C directories. Also check out the [Testing Documentation](http://www.sublimetext.com/docs/3/syntax.html#testing).

**Comment by wbond** [comment]
This should be resolved with b80aa26609c389ae1b9901da477e2d05cedaab6b
